crypto/sm2/sm2_crypt.c

```diff
@@ -391,3 +391,350 @@ int sm2_decrypt(const EC_KEY *key,
+
+    if (!out || !outlen) return retval;
+    if (md == NULL) md = EVP_sm3();
+    mdlen = EVP_MD_size(md);
+    mctx = EVP_MD_CTX_new();
+    if (mctx == NULL) {
+        SM2err(SM2_F_KDF_GMT003_2012, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+
+    for (counter = 1;; counter++)
+    {
+        unsigned char dgst[EVP_MAX_MD_SIZE];
+
+        EVP_DigestInit(mctx, md);
+        ctr[0] = (unsigned char)((counter >> 24) & 0xFF);
+        ctr[1] = (unsigned char)((counter >> 16) & 0xFF);
+        ctr[2] = (unsigned char)((counter >> 8) & 0xFF);
+        ctr[3] = (unsigned char)(counter & 0xFF);
+        if (!EVP_DigestUpdate(mctx, Z, Zlen))
+            goto err;
+        if (!EVP_DigestUpdate(mctx, ctr, sizeof(ctr)))
+            goto err;
+        if (!EVP_DigestUpdate(mctx, SharedInfo, SharedInfolen))
+            goto err;
+        if (!EVP_DigestFinal(mctx, dgst, NULL))
+            goto err;
+
+        if (outlen > mdlen)
+        {
+            memcpy(out, dgst, mdlen);
+            out += mdlen;
+            outlen -= mdlen;
+        }
+        else
+        {
+            memcpy(out, dgst, outlen);
+            memset(dgst, 0, mdlen);
+            break;
+        }
+    }
+
+    retval = 1;
+
+err:
+    EVP_MD_CTX_free(mctx);
+    return retval;
+}
+
+
+int SM2Kap_compute_key(void *out, size_t outlen, int server,\
+    const char *peer_uid, int peer_uid_len, const char *self_uid, int self_uid_len, \
+    const EC_KEY *peer_ecdhe_key, const EC_KEY *self_ecdhe_key, const EC_KEY *peer_pub_key, const EC_KEY *self_eckey, \
+    const EVP_MD *md)
+{
+    BN_CTX *ctx = NULL;
+    EC_POINT *UorV = NULL;
+    const EC_POINT *Rs, *Rp;
+    BIGNUM *Xs = NULL, *Xp = NULL, *h = NULL, *t = NULL, *two_power_w = NULL, *order = NULL;
+    const BIGNUM *priv_key, *r;
+    const EC_GROUP *group;
+    int w;
+    int ret = -1;
+    size_t buflen, len;
+    unsigned char *buf = NULL;
+
+    if (outlen > INT_MAX)
+    {
+        SM2err(SM2_F_SM2KAP_COMPUTE_KEY, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+
+    if (!peer_pub_key || !self_eckey)
+    {
+        SM2err(SM2_F_SM2KAP_COMPUTE_KEY, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    
+    priv_key = EC_KEY_get0_private_key(self_eckey);
+    if (!priv_key)
+    {
+        SM2err(SM2_F_SM2KAP_COMPUTE_KEY, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+
+    if (!peer_ecdhe_key || !self_ecdhe_key)
+    {
+        SM2err(SM2_F_SM2KAP_COMPUTE_KEY, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+
+    Rs = EC_KEY_get0_public_key(self_ecdhe_key);
+    Rp = EC_KEY_get0_public_key(peer_ecdhe_key);
+    r = EC_KEY_get0_private_key(self_ecdhe_key);
+
+    if (!Rs || !Rp || !r)
+    {
+        SM2err(SM2_F_SM2KAP_COMPUTE_KEY, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+
+    ctx = BN_CTX_new();
+    Xs = BN_new();
+    Xp = BN_new();
+    h = BN_new();
+    t = BN_new();
+    two_power_w = BN_new();
+    order = BN_new();
+
+    if (!Xs || !Xp || !h || !t || !two_power_w || !order)
```

Comment:
Only the last BN from the context needs to be checked.  I.e. `if (order == NULL)`